torch/fx/experimental/Partitioner.py

```diff
@@ -0,0 +1,165 @@
+        for child in self.children:
+            line +=  'partition_' + str(child) + ' '
+        line += '\n'
+        line += 'input nodes: '
+        for node in self.input_nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        line += '\n'
+        line += 'output nodes: '
+        output_nodes = self.output_nodes
+        for node in output_nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        return line
+
+"""DAG class contains all the DAG nodes"""
+class DAG:
+    def __init__(self) -> None:
+        self.nodes = []
+    def create_node(self, partition_id:int, parents_ids: List[int], children_ids: List[int], input_nodes: List[Node], output_nodes: List[Node]) -> None:
+        node = DAG_node(partition_id, parents_ids, children_ids, input_nodes, output_nodes)
+        self.nodes.append(node)
+
+"""Partition class contains all the information about an individual partition.
+   It also provides necessary methods for manipulation the partition.
+"""
+class Partition:
+    def __init__(self, partition_id: int, fx_module: GraphModule) -> None:
+        self.graph_module = fx_module
+        self.nodes = []
+        self.partition_id = partition_id
+        self.parents = []
+        self.children = []
+
+    def add_node(self, node: Node) -> None:
+        """Append a new node into the partition."""
+        self.nodes.append(node)
+
+    def add_parent(self, partition: 'Partition') -> None:
+        self.parents.append(partition)
+
+    def add_child(self, partition: 'Partition') -> None:
+        self.children.append(partition)
+
+    def get_children(self) -> List['Partition']:
+        return self.children
+
+    def get_parents(self) -> List['Partition']:
+        return self.parents
+
+    def get_input_nodes(self) -> List[Node]:
+        """Input nodes are coming from two places:
+        placeholder and output from its parents output.
+        """
+        input_nodes: List[Node] = []
+        for node in self.nodes:
+            if node.op == 'placeholder':
+                input_nodes.append(node)
+        for partition in self.parents:
+            input_nodes += partition.get_output_nodes()
+        return input_nodes
+
+    def get_output_nodes(self) -> List[Node]:
+        """Output nodes are the nodes that without any user inside this partition."""
+        output_nodes: List[Node] = []
+        for node in self.nodes:
+            index = self.graph_module.graph.nodes.index(node)
+            user_indexes = GraphManipulation.get_all_users_of(self.graph_module, index)
+            user_nodes = set([self.graph_module.graph.nodes[i] for i in user_indexes])
+            # check if user nodes has an intersection with self.nodes
+            if not set(self.nodes).intersection(user_nodes):
+                output_nodes.append(node)
+        return output_nodes
+
+    def __str__(self) -> str:
+        return str(self.partition_id)
+
+"""A graph module may not fit into one device.
```

Comment:
General nit: The docstring for a class should be right after the definition, same goes for functions.
https://www.programiz.com/python-programming/docstrings has a good discussion of docstring style.